slatedb/src/garbage_collector/compacted_gc.rs

```diff
@@ -77,55 +111,63 @@ impl CompactedGcTask {
+        // manifest to object storage).
+        let compaction_low_watermark_dt = self.compaction_low_watermark_dt();
+        let active_manifests = self.manifest_store.read_active_manifests().await?;
+        let active_ssts = self
+            .list_active_l0_and_compacted_ssts(&active_manifests)
+            .await?;
+        // Don't delete any SSTs that are newer than the configured minimum age.
+        let configured_min_age_dt = utc_now - self.compacted_sst_min_age();
+        // Don't delete SSTs that are newer than this SST since they're probably an L0 that hasn't yet
+        // been added to the manifest (we write the L0, _then_ add it to the manifest and write the
+        // manifest to object storage).
+        let most_recent_sst_dt = self.newest_l0_dt(&active_manifests).await?;
+        // Take the minimum of the configured min age, the compaction low watermark, and the most
+        // recent SST in the manifest. This is the true upper-limit for SSTs that may be deleted.
+        let cutoff_dt = configured_min_age_dt
+            .min(compaction_low_watermark_dt)
+            .min(most_recent_sst_dt);
+        log::debug!(
+            "calculated compacted SST GC cutoff [cutoff_dt={:?}, configured_min_age_dt={:?}, compaction_low_watermark_dt={:?}, most_recent_sst_dt={:?}]",
+            cutoff_dt,
+            configured_min_age_dt,
+            compaction_low_watermark_dt,
+            most_recent_sst_dt,
+        );
         let sst_ids_to_delete = self
             .table_store
             // List all SSTs in the table store
             .list_compacted_ssts(..)
             .await?
             .into_iter()
-            // Filter out the ones that are too young to be collected
-            .filter(|sst| utc_now.signed_duration_since(sst.last_modified) > min_age)
             .map(|sst| sst.id)
-            // Filter out the ones that are active in the manifest
+            // Filter out SSTs that were more recently created than the cutoff_dt
+            .filter(|id| DateTime::<Utc>::from(id.unwrap_compacted_id().datetime()) < cutoff_dt)
+            // Filter out SSTs that are active in the manifest (including actively checkpointed SSTs)
             .filter(|id| !active_ssts.contains(id))
-            // Filter out the ones that may be part of a running compaction (See #604 for details)
-            .filter(|id| match (maybe_compaction_low_watermark_dt, id) {
-                (Some(compaction_low_watermark_dt), SsTableId::Compacted(ulid)) => {
-                    let ulid_dt = DateTime::<Utc>::from(ulid.datetime());
-                    ulid_dt < compaction_low_watermark_dt
-                }
-                // Keep all compacted SSTs if there is no compaction low watermark
-                (None, SsTableId::Compacted(_)) => true,
-                _ => unreachable!("non-compacted SSTs should not be in list_compacted_ssts"),
-            })
             .collect::<Vec<_>>();
 
         for id in sst_ids_to_delete {
```

Comment:
Its probably worth writing an info log whenever we delete an SST. If we do that we can also add the following to `db_state.rs` to get the debug logger to print the sst IDs using their string representation (the default `Debug` impl just dumps the raw u128)

```
impl Debug for SsTableId {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
        match self {
            Wal(id) => write!(f, "SsTableId({})", id),
            Compacted(id) => write!(f, "SsTableId({})", id.to_string()),
        }
    }
}
```